src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenNullCheckedParameterTests.cs

```diff
@@ -0,0 +1,456 @@
+    IL_0009:  ret
+}");
+            // Debug
+            compilation = CompileAndVerify(source, options: TestOptions.DebugExe);
+            compilation.VerifyIL("C.M", @"
+{
+    // Code size       12 (0xc)
+    .maxstack  3
+    IL_0000:  nop
+    IL_0001:  ldarg.0
+    IL_0002:  ldc.i4.1
+    IL_0003:  ldc.i4.s   77
+    IL_0005:  call       ""void C.M2(__arglist) with __arglist( int, char)""
+    IL_000a:  nop
+    IL_000b:  ret
+}");
+        }
+
+        [Fact]
+        public void TestManyNullCheckedArgs()
+        {
+            var source = @"
+class C
+{
+    public void M(int x!, int y!) { }
+    public static void Main() { }
+}";
+
+            // Release
+            var compilation = CompileAndVerify(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyIL("C.M(int, int)", @"
+{
+  // Code size       19 (0x13)
+  .maxstack  1
+ ~IL_0000:  ldarg.1
+  IL_0001:  brtrue.s   IL_0009
+  IL_0003:  newobj     ""System.Exception..ctor()""
+  IL_0008:  throw
+ ~IL_0009:  ldarg.2
+  IL_000a:  brtrue.s   IL_0012
+  IL_000c:  newobj     ""System.Exception..ctor()""
+  IL_0011:  throw
+ -IL_0012:  ret
+}", sequencePoints: "C.M");
+            // Debug
+            compilation = CompileAndVerify(source, options: TestOptions.DebugExe);
+            compilation.VerifyIL("C.M(int, int)", @"
+  {
+  // Code size       20 (0x14)
+  .maxstack  1
+ ~IL_0000:  ldarg.1
+  IL_0001:  brtrue.s   IL_0009
+  IL_0003:  newobj     ""System.Exception..ctor()""
+ IL_0008:  throw
+ ~IL_0009:  ldarg.2
+  IL_000a:  brtrue.s   IL_0012
+  IL_000c:  newobj     ""System.Exception..ctor()""
+  IL_0011:  throw
+ -IL_0012:  nop
+ -IL_0013:  ret
+  }", sequencePoints: "C.M");
+        }
+
+        [Fact]
+        public void TestNullCheckedIndexedProperty()
+        {
+            var source = @"
+class C
+{
+    public string this[int index!] => null;
+    public static void Main() { }
+}";
+
+            // Release
+            var compilation = CompileAndVerify(source, options: TestOptions.ReleaseExe);
+            compilation.VerifyIL("C.this[int].get", @"
+{
+  // Code size       11 (0xb)
+  .maxstack  1
+ ~IL_0000:  ldarg.1
+  IL_0001:  brtrue.s   IL_0009
+  IL_0003:  newobj     ""System.Exception..ctor()""
+  IL_0008:  throw
+ -IL_0009:  ldnull
+  IL_000a:  ret
+}", sequencePoints: "C.get_Item");
+
+            // Debug
+            compilation = CompileAndVerify(source, options: TestOptions.DebugExe);
+            compilation.VerifyIL("C.this[int].get", @"
+{
+   // Code size       11 (0xb)
+   .maxstack  1
+  ~IL_0000:  ldarg.1
+   IL_0001:  brtrue.s   IL_0009
+   IL_0003:  newobj     ""System.Exception..ctor()""
+   IL_0008:  throw
+  -IL_0009:  ldnull
+   IL_000a:  ret
+}", sequencePoints: "C.get_Item");
+        }
+
+        [Fact]
+        public void TestNullCheckedIndexedGetterSetter()
+        {
+            var source = @"
+class C
+{
+    private string[] names;
+    public string this[int index!]
```

Comment:
>int [](start = 23, length = 3)

Consider testing `object index!` instead. (There's already a test for `int index!` above.)